### PR TITLE
Disalow to set arbitrary environments for plugins

### PR DESCRIPTION
**Pull request by 6543**
also split the different step types in our json schema to lint accordingly

### CODE REVIEW

File: pipeline/frontend/yaml/linter/linter.go [hunk 1]
```diff
@@ -133,15 +133,14 @@ func (l *Linter) lintImage(config *WorkflowConfig, c *types.Container, area stri
...
-		var keys []string
-		for key := range c.Settings {
-			keys = append(keys, key)
-		}
-		return newLinterError(fmt.Sprintf("Cannot configure both commands and custom attributes %v", keys), config.File, fmt.Sprintf("%s.%s", field, c.Name), false)
+	if len(c.Settings) != 0 && len(c.Commands) != 0 {
+		return newLinterError("Cannot configure both commands and settings", config.File, fmt.Sprintf("%s.%s", field, c.Name), false)
+	}
+	if len(c.Settings) != 0 && len(c.Environment) != 0 {
+		return newLinterError("Cannot configure both commands and settings", config.File, fmt.Sprintf("%s.%s", field, c.Name), false)
```
Comment (6543) [inline pipeline/frontend/yaml/linter/linter.go:143]:
```suggestion
		return newLinterError("Cannot configure both environment and settings", config.File, fmt.Sprintf("%s.%s", field, c.Name), false)
```

File: pipeline/frontend/yaml/linter/linter.go [hunk 2]
```diff
@@ -133,15 +133,14 @@ func (l *Linter) lintImage(config *WorkflowConfig, c *types.Container, area stri
...
-	if len(c.Settings) != 0 {
-		var keys []string
-		for key := range c.Settings {
-			keys = append(keys, key)
-		}
-		return newLinterError(fmt.Sprintf("Cannot configure both commands and custom attributes %v", keys), config.File, fmt.Sprintf("%s.%s", field, c.Name), false)
+	if len(c.Settings) != 0 && len(c.Commands) != 0 {
+		return newLinterError("Cannot configure both commands and settings", config.File, fmt.Sprintf("%s.%s", field, c.Name), false)
+	}
+	if len(c.Settings) != 0 && len(c.Environment) != 0 {
```
Comment (6543) [inline pipeline/frontend/yaml/linter/linter.go:142]: Entrypoint is missing ... and we should cover it witj tests

File: pipeline/frontend/yaml/linter/schema/schema.json [hunk 3]
```diff
@@ -304,10 +304,24 @@
       }
     },
     "step": {
+      "description": "An step of your workflow executes either arbitrary commands or uses an plugin. Read more: https://woodpecker-ci.org/docs/usage/workflow-syntax#steps",
```
Comment (xoxys) [inline pipeline/frontend/yaml/linter/schema/schema.json:307]:
```suggestion
      "description": "A step of your workflow executes either arbitrary commands or uses a plugin. Read more: https://woodpecker-ci.org/docs/usage/workflow-syntax#steps",
```

**Comment by 6543** [comment]
ok I did go throug all the documentation and using `environment` in combination with `settings` was never showcased once and never mentioned either.

The plugin documentation is a bit lacking and can be improved.

## but based on our consensus to count changes of features we never have documented as non breaking, this is non breaking! :tada:

**Comment by wez** [comment]
FWIW, this broke my pipeline, do you have suggestions on how to migrate this to a working configuration?
I need to set both environment and settings to workaround a drone compatibility issue:

```yaml
merge-docker-image-arm:
  depends_on: [build-docker-image-arm]
  when:
    - evaluate: 'platform == "linux/aarch64" && UBUNTU_IMAGE == "ubuntu:22.04" && CI_PIPELINE_EVENT == "tag"'
  image: plugins/manifest
  failure: ignore
  environment: &docker_manifest_env
    # Legacy env var to prevent the plugin from throwing an error
    # when converting an empty string to a number
    PULLREQUEST_DRONE_PULL_REQUEST: 0
  settings: &docker_manifest_settings_tag
    <<: *docker_credentials
    target: "something"
    template: "something:${CI_COMMIT_TAG}-ARCH"
    tags:
      - latest
      - ${CI_COMMIT_TAG}
    platforms:
      - linux/amd64
      - linux/arm64
```

**Comment by 6543** [comment]
uh that should land in our compatibitity layer instead!

I'll send a pull and you can use that and drop your workaround

**Comment by 6543** [comment]
# If you encounter an similar issue please open a new issue and only link this pull :)

_locking now_

**Pull request merged**